calico.go

```diff
@@ -0,0 +1,265 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"runtime"
+
+	"github.com/vishvananda/netlink"
+
+	"flag"
+
+	"github.com/containernetworking/cni/pkg/ip"
+	"github.com/containernetworking/cni/pkg/ipam"
+	"github.com/containernetworking/cni/pkg/ns"
+	"github.com/containernetworking/cni/pkg/skel"
+	"github.com/containernetworking/cni/pkg/types"
+	"github.com/projectcalico/calico-cni/k8s"
+	. "github.com/projectcalico/calico-cni/utils"
+	"github.com/projectcalico/libcalico/lib"
+)
+
+var hostname string
+
+func init() {
+	// This ensures that main runs only on main thread (thread group leader).
+	// since namespace ops (unshare, setns) are done for a single thread, we
+	// must ensure that the goroutine does not jump from OS thread to thread
+	runtime.LockOSThread()
+
+	hostname, _ = os.Hostname()
+}
+
+func cmdAdd(args *skel.CmdArgs) error {
+	if err := AddIgnoreUnknownArgs(); err != nil {
+		return err
+	}
+
+	// Unmarshall the network config, and perform validation
+	conf := NetConf{}
+	if err := json.Unmarshal(args.StdinData, &conf); err != nil {
+		return fmt.Errorf("failed to load netconf: %v", err)
+	}
+
+	// Allow the hostname to be overridden by the network config
+	if conf.Hostname != "" {
+		hostname = conf.Hostname
+	}
+
+	if err := ValidateNetworkName(conf.Name); err != nil {
+		return err
+	}
+
+	etcd, err := libcalico.GetKeysAPI(conf.EtcdAuthority, conf.EtcdEndpoints)
+	if err != nil {
+		return err
+	}
+
+	workloadID, orchestratorID, err := GetIdentifiers(args)
+	if err != nil {
+		return err
+	}
+
+	// Always check if there's an existing endpoint.
+	endpoint, err := libcalico.GetEndpoint(
+		etcd, libcalico.Workload{
+			Hostname:       hostname,
+			OrchestratorID: orchestratorID,
+			WorkloadID:     workloadID})
+
+	if err != nil {
+		return err
+	}
+
+	fmt.Fprintf(os.Stderr, "Calico CNI checking for existing endpoint. endpoint=%v\n", endpoint)
+
+	// Collect the result in this variable - this is ultimately what gets "returned" by this function by printing
+	// it to stdout.
+	var result *types.Result
+
+	// If running under Kubernetes then branch off into the kubernetes code, otherwise handle everything in this
+	// function.
+	if orchestratorID == "k8s" {
+		if result, err = k8s.CmdAddK8s(args, conf, hostname, etcd, endpoint); err != nil {
+			return err
+		}
+	} else {
+		// Not running in an environment that needs special casing.
+		profileID := conf.Name
+
+		if endpoint != nil {
+			// Don't create the veth or do any networking. Just update the profile on the endpoint. The profile will
+			// be created if needed during the profile processing step.
+
```

Comment:
nit: I think the first bit of comment in this branch should be "There is an existing endpoint - no need to create another.  This occurs when adding an existing container to a new CNI network" or something similar so that it is abundantly obvious what case this branch is handling.  The rest of these comments which explain what we actually do in this branch.